### ISSUE TITLE
Allow better control over weel tags

### ISSUE THREAD

**Issue by FFY00** [mesonbuild/meson-python#639]
https://github.com/pypackaging-native/pkgconf-pypi produces platform-specific but Python-independent tags. RIght now, I don't think there's any way to expose this.

**Comment by rgommers** [comment]
IIRC this should work. Just tested with https://github.com/mesonbuild/meson-python/tree/main/tests/packages/executable, which yields:
```
executable-1.0.0-py3-none-linux_x86_64.whl
```
What would you expect if not those tags?

**Comment by FFY00** [comment]
Yes, but the code makes some assumptions that aren't true for our project.

https://github.com/mesonbuild/meson-python/blob/063fe8bc8ad6984c29b24a0f72a14e1ced091a9c/mesonpy/__init__.py#L319-L349

Specifically, it assumes that we are not installing platform binary files to the Python directory.

**Comment by rgommers** [comment]
IIRC the `ninja` wheel has a very similar structure.

**Comment by FFY00** [comment]
> Got it, thanks. Yes, we are trying to infer tags and that isn't correct in this case. It's actually not so easy to always be 100% correct here, which is why some other build backends have an override mechanism to force the tags to be a certain value.

Yes, I think the best you we can do right now would be looking at the meson tags, but I think it'd be better to also have the option to override.

**Comment by ncoghlan** [comment]
A different-but-related use case: when building on later Python versions with `Py_LIMITED_API` set to the limit the symbols to those published by an earlier version, I wasn't able to find a way to tell meson to set the Python tag to `cp311` (the declared API version) instead of to `cp312` (the build environment).

The workaround was straightforward (just rename the file afterwards), but it seemed worth mentioning as a use case if improved native support is considered in the future.

**Comment by eli-schwartz** [comment]
What would be a nice-to-have is if there was some way to analyze a compiled extension module and determine what is the minimum ABI version it uses.

glibc does something somewhat similar-ish. The "stable ABI" is libc.so.6, and looking at what symbol versions it uses of the form `@GLIBC_2.XX` tells you what the minimum glibc version it is compatible is. This can then be reported by `auditwheel repair`:

```
The wheel references external versioned symbols in these
system-provided shared libraries: libc.so.6 with versions
{'GLIBC_2.4', 'GLIBC_2.14', 'GLIBC_2.2.5'}

This constrains the platform tag to "manylinux_2_17_x86_64". In order
to achieve a more compatible tag, you would need to recompile a new
wheel from source on a system with earlier versions of these
libraries, such as a recent manylinux image.
```

**Comment by rgommers** [comment]
In gh-842 there was another case (only a `library()`, that gets mistaken for an extension module) that would be helped by either better control over wheel tags, or better inference. 

In principle we can detect whether a `.so` is an extension module yes or no, by checking for a `PyInit_` symbol (and whatever is needed for [PEP 793](https://peps.python.org/pep-0793/) after `PyInit_` disappears). We only have to find a single extension module, so it wouldn't be a performance concern I think. 

An override might be better perhaps. Adding that also doesn't stop improving the inference, both might be useful.

**Comment by eli-schwartz** [comment]
> In principle we can detect whether a `.so` is an extension module yes or no, by checking for a `PyInit_` symbol (and whatever is needed for [PEP 793](https://peps.python.org/pep-0793/) after `PyInit_` disappears). We only have to find a single extension module, so it wouldn't be a performance concern I think.

We can sort of detect this purely from meson-info/intro-targets.json, as python extension modules get subclassed from shared_module (they are plugin modules that operate dlopen-style, so this is technically correct) and this information is in the introspection data as the "type" attribute.

We could also perhaps enhance the introspection to have a dedicated "python_module" type, which would allow greater accuracy. It's probably "just" a slightly messy backend refactor.

A python module will never be of type library / shared_library. Those have sonames and are intended to be linked to, not dlopened. Introspection  is already good enough to tell you that today (even though you cannot just guess via the filename, since knowing it ends in .so doesn't tell you whether or not it's an extension module).
